pkg/collector/agonsticapi/check.go

```diff
@@ -0,0 +1,289 @@
+	Metrics []metric `json:"metrics"`
+}
+
+type agnosticCheck struct {
+	sendermanager sender.SenderManager
+	tagger        tagger.Component
+	name          string
+	libHandle     unsafe.Pointer
+	id            checkid.ID
+}
+
+func newCheck(sendermanager sender.SenderManager, tagger tagger.Component, name string, lib unsafe.Pointer) (check.Check, error) {
+	return &agnosticCheck{
+		sendermanager: sendermanager,
+		tagger:        tagger,
+		name:          name,
+		libHandle:     lib,
+	}, nil
+}
+
+func (c *agnosticCheck) Run() error {
+	var cErr *C.char
+
+	result := C.run_agnostic_check(c.libHandle, &cErr)
+
+	if cErr != nil {
+		errorString := C.GoString(cErr)
+		defer C.free(unsafe.Pointer(cErr))
+
+		errMsg := fmt.Sprintf("failed to execute `run`  function on library %s: err %s", c.name, errorString)
+		return errors.New(errMsg)
+	}
+
+	if result == nil {
+		return fmt.Errorf("library %s execution did not returned any result", c.name)
+	}
+
+	// Extract the JSON string from the Result using the len attribute
+	jsonString := C.GoStringN(result.Char, result.Len)
+
+	var payload payload
+	if err := json.Unmarshal([]byte(jsonString), &payload); err != nil {
+		return fmt.Errorf("failed to unmarshal JSON from library %s: %v", c.name, err)
+	}
+
+	C.free_result(c.libHandle, result, &cErr)
+	if cErr != nil {
+		errorString := C.GoString(cErr)
+		defer C.free(unsafe.Pointer(cErr))
+		// Log the error but don't fail the check since we already got the data
+		fmt.Printf("Warning: failed to free result from library %s: %s\n", c.name, errorString)
+	}
+
+	fmt.Printf("Received %d metrics from library %s\n", len(payload.Metrics), c.name)
+	for _, metric := range payload.Metrics {
+		fmt.Printf("Metric: %s = %f (type: %s, tags: %v)\n", metric.Name, metric.Value, metric.Type, metric.Tags)
+		sender, err := c.sendermanager.GetSender(c.id)
+		if err != nil {
+			return fmt.Errorf("failed to get sender manager for shared library check %s: %v", c.name, err)
+		}
+
+		name := metric.Name
+		value := metric.Value
+		hostn, _ := hostname.Get(context.TODO())
+		tags := metric.Tags
+
+		// We need to figure out what tags we need to fetch from the Agent
+		extraTags, err := c.tagger.GlobalTags(types.LowCardinality)
+		if err != nil {
+			fmt.Printf("failed to get global tags: %s\n", err.Error())
+		} else {
+			tags = append(tags, extraTags...)
+		}
+
+		switch metric.Type {
+		case "gauge":
+			sender.Gauge(name, value, hostn, tags)
+		case "rate":
+			sender.Rate(name, value, hostn, tags)
+		case "count":
+			sender.Count(name, value, hostn, tags)
+		case "monotonic_count":
+			sender.MonotonicCountWithFlushFirstValue(name, value, hostn, tags, false)
+		case "counter":
+			sender.Counter(name, value, hostn, tags)
+		case "histogram":
+			sender.Histogram(name, value, hostn, tags)
+		case "historate":
+			sender.Historate(name, value, hostn, tags)
+		}
+	}
+
+	return nil
+}
+
+func (c *agnosticCheck) Stop() {}
+
+func (c *agnosticCheck) Cancel() {
```

Comment:
Is the agent expected to call `Cancel` on each dynamically loaded check on shutdown, or do we just exit and let the OS figure it out? If the former, renaming the API to something more intuitive (`Unload` or `Close` would make more sense to me) might be helpful.